Switch provider-template to use the Upbound build submodule

negz opened the pull request:
I find that in practice I personally change every new repo over to use this. All or most of the providers under the crossplane and crossplane-contrib orgs do the same. I've tested this by running `make reviewable test`.

I've tested this by running `make && make reviewable` on a fresh clone of the repo.

hasheddan reviewed (commented):
I don't _personally_ like adding the build submodule since it feels like we mostly want it when a repo is created in the Crossplane or Crossplane contrib org, in which case someone who is knowledgeable of the submodule is likely bootstrapping the project and could easily add it. However, I'm not really sure how often this template is being utilized outside of this use case, so maybe it makes sense to add it. @negz it sounds like @krishchow is planning to work on some scaffolding functionality in the CLI with https://github.com/crossplane/crossplane/issues/2188 so maybe that could be used for the "lightweight" case?

negz commented:
Do we have an idea of how many providers that are in active use are not using the build submodule? Would concerns about the submodule be alleviated if we addressed https://github.com/crossplane/crossplane/issues/1583 (CC @AaronME)

hasheddan commented:
@negz I am guessing very few. I do think even if we fork the build submodule it is still pretty opaque for contributors, but maybe with the functionality it provides that is somewhat unavoidable

negz commented:
@benagricola @chlunde @cheewaio It sounds like you've all been using provider-template recently. What do you think about this change? Do you typically use the build submodule?

benagricola commented:
> @benagricola @chlunde @cheewaio It sounds like you've all been using provider-template recently. What do you think about this change? Do you typically use the build submodule?

I've not had any issue using the build submodule in the projects it's already in, but I do agree it is somewhat opaque, from the few instances where I've had to work out how a build job is functioning.

I've not had any issue with the existing `Makefile` in `provider-template` either and it is definitely clearer, although I did sorta assume that the `build` submodule was used here too and was a bit surprised when `make reviewable` didn't work (although the equivalent commands still exist).

I suppose my opinion is that the existing `Makefile` is fine for my purposes right now, in lieu of any particular reason to use the `build` submodule based on the additional complexity / opaqueness - however maybe this is different to others' experiences in that I templated to my own repo which has no automated build process yet (I don't know if the build submodule adds anything here?)

chlunde commented:
> It sounds like you've all been using provider-template recently. What do you think about this change? Do you typically use the build submodule?

I don't know what I would gain with the build submodule, skimming the submodule it's not obvious what's used except to pin the kubectl version? Personally I think it might be nice with a template for github actions (perhaps using github as a registry, if that's easiest) and release conventions.

AaronME commented:
> Do we have an idea of how many providers that are in active use are not using the build submodule? (CC @AaronME)

Based on a quick search, the following crossplane project are including the build submodule:
- crossplane-contrib/agent
- crossplane-contrib/provider-argocd
- crossplane-contrib/provider-capi
- crossplane-contrib/provider-cloudinit
- crossplane-contrib/provider-digitalocean
- crossplane-contrib/provider-equinix
- crossplane-contrib/provider-equinix-metal
- crossplane-contrib/provider-github
- crossplane-contrib/provider-gitlab
- crossplane-contrib/provider-helm
- crossplane-contrib/provider-ibm-cloud
- crossplane-contrib/provider-in-cluster
- crossplane-contrib/provider-nop
- crossplane-contrib/provider-secret
- crossplane-contrib/provider-sql
- crossplane-contrib/provider-terraform-vsphere
- crossplane/addon-oam-kubernetes-remote
- crossplane/conformance
- crossplane/crossplane
- crossplane/crossplane-runtime
- crossplane/crossplane-tools
- crossplane/gitlab-controller
- crossplane/oam-kubernetes-runtime
- crossplane/pkg-unpack
- crossplane/provider-alibaba
- crossplane/provider-aws
- crossplane/provider-azure
- crossplane/provider-gcp
- crossplane/provider-gcp-beta
- crossplane/provider-rook
- crossplane/release-test
- crossplane/templating-controller

chlunde commented:
I'm now in favor of doing this, because having CI working out of the box is nice. @larhauga and @olesu added this module to provider-bitbucket-server after I wrote the initial code, and they discovered lot's of missing files and lint issues... I think it's easier to remove from projects that want a different build system later.

Ref.: https://github.com/crossplane-contrib/provider-bitbucket-server/pull/12

muvaf commented:
@negz Are you still interested in merging this change? If you are, it'd be great to merge it and tackle updating the documentation in https://github.com/crossplane/crossplane/issues/2737

negz commented:
@muvaf I am - I'll add it to my stack for this week.

negz commented:
I've rebased and updated this. It seems to work (at least `make && make reviewable` does). I'd like to add all the usual CI/CD workflows before I move it out of draft though.

negz commented:
This is now ready to go.

muvaf commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,40 @@
+labels: bug
+---
+<!--
+Thank you for helping to improve Crossplane!
```
Would it make sense to `s/Crossplane/Provider Template/g` in these files?

negz replied:
I don't think so - we use "Crossplane" in the broader project sense for most provider repos that already have this issue template.

muvaf commented on .github/ISSUE_TEMPLATE/release.md:
```diff
@@ -0,0 +1,125 @@
+- [ ] Updated the current release version in the [Crossplane docs website repo] ([link][website-pr]).
+- [ ] Updated the release branch reaching EOL with docs removal directive ([link][eol-pr]).
+-->
+- [ ] Updated all version information in the documentation on the relevant release branch ([link][docs-update-pr]).
```
Seems to list tasks for a Crossplane rather than provider release.

stevendborrelli replied:
This should be updated to the provider-template specific workflow.

negz replied:
I added this template unintentionally - I don't think we should include it at all. Most providers don't currently have such an issue template.

muvaf commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,324 @@
+on:
+  push:
+    branches:
+      - master
```
```suggestion
      - main
```

negz replied:
Good catch - will fix.

muvaf commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,42 @@
+on:
+  push:
+    branches:
+      - master
```
```suggestion
      - main
```

muvaf commented on .github/workflows/promote.yml:
```diff
@@ -0,0 +1,48 @@
+  # a step 'if env.AWS_USR' != ""', so we copy these to succinctly test whether
+  # credentials have been provided before trying to run steps that need them.
+  DOCKER_USR: ${{ secrets.DOCKER_USR }}
+  AWS_USR: ${{ secrets.AWS_USR }}
```
```suggestion
```
I don't think we need AWS user for providers since they don't publish an artifact other than a Docker image.

negz replied:
You're probably right, but I kind of want to keep this as is. Our general pattern is to pretty much try keep the CI file the same everywhere so we can just copy and paste to propagate new versions with minimal updates.

muvaf commented on Makefile:
```diff
@@ -1,42 +1,107 @@
-	kubectl apply -f package/crds/ -R
-	go run cmd/provider/main.go -d
+# Setup Images
+DOCKER_REGISTRY ?= crossplane
```
We have used this so far. Do you think we should keep using that org? I can see us pushing `crossplane-contrib` images to a `crossplane-contrib` org.

negz replied:
I'd prefer to mostly propagate what we do today in this PR and make changes in future PRs. So let's keep this for now.

muvaf commented on cluster/images/provider-template-controller/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM BASEIMAGE
```
Should we make the same changes from https://github.com/crossplane/crossplane/pull/2718 ?

stevendborrelli replied:
I have a PR https://github.com/crossplane-contrib/provider-kafka/pull/20 that migrates another provider-template based project onto the new makefile. Perhaps we can use it as a template?

negz replied:
I'd prefer to mostly propagate what we do today in this PR and make changes in future PRs. So let's keep this for now.

muvaf commented on package/crds/sample.template.crossplane.io_mytypes.yaml:
```diff
@@ -1,3 +1,5 @@
+
+---
```
These two lines should've been cleaned actually.

stevendborrelli commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,324 @@
+  # Common versions
+  GO_VERSION: '1.16'
+  GOLANGCI_VERSION: 'v1.31'
+  DOCKER_BUILDX_VERSION: 'v0.4.2'
```
v0.4.2 is 1+ year old, should we update buildx? The current version is 0.7.1, My Docker Desktop on a Mac M1 is showing v0.6.3.

negz replied:
Let's do that in a follow up. I'd prefer to keep this PR mostly scoped to copying the config we're using everywhere else. (Keeping that config updated across all our repos is another problem.)

stevendborrelli reviewed (approved):
Apart from a few nits, this looks good. We can update it to use the `imagelight.mk` if not in this PR with another one shortly after.

negz reviewed (commented):
I pushed some commits that I believe address all of the feedback (at least, everything I think is blocking this PR).

hasheddan commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,324 @@
+
+      - name: Find the Go Build Cache
+        id: go
+        run: echo "::set-output name=cache::$(go env GOCACHE)"
```
@negz we'll need to update all of these to ensure that we actually identify the correct cache
```suggestion
        run: echo "::set-output name=cache::$(make go.cachedir)"
```

negz replied:
Thanks - fixed.
